Stack calculator

MinyazevR opened the pull request.

MinyazevR committed:
Writing basic stack functions, writing tests

MinyazevR committed:
Rollback of commits due to their uninformativeness

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using Stack;
+
+namespace StackCalculator;
```
namespace лучше в самом начале писать

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+/// </summary>
+public class Calculator
+{
+    private IStack<float> Stack = new StackOnLists<float>();
```
По условию, ` в коде класса «Стековый калькулятор» не должно быть ни одного упоминания конкретных реализаций стека, даже если очень хочется`.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+    /// <returns>Expression value</returns>
+    public float CountTheExpressionInPostfixForm(string[] inputString)
+    {
+        int number = 0;
```
Слишком рано объявлен. Надо его объявить в месте инициализации: `if (int.TryParse(inputString[i], out int number))`

yurii-litvinov commented on Stack/Stack/IStack.cs:
```diff
@@ -0,0 +1,47 @@
+/// <summary>
+/// A class representing the stack interface
+/// </summary>
+abstract public class IStack<T>
```
Это должен был быть интерфейс, а не абстрактный класс

yurii-litvinov commented on Stack/Stack/IStack.cs:
```diff
@@ -0,0 +1,47 @@
+    /// Function that returns the number of elements in the stack
+    /// </summary>
+    /// <returns>Number of elements in stack</returns>
+    abstract public int ReturnNumberOfElements();
```
Идеологически правильнее `NumberOfElements();`, "Return" тут несколько избыточно.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+            {
+                if (Stack.ReturnNumberOfElements() < 2)
+                {
+                    throw new IncorrectExpressionException("Incorrect expression");
```
То, что Incorrect expression, понятно из типа исключения, тут Message не нужен

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+                    throw new IncorrectExpressionException("Incorrect expression");
+                }
+            }
+            catch (IncorrectExpressionException exception)
```
Делается throw и тут же catch, то есть то самое, что я на паре говорил не делать :) Это бессмысленный код.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+            }
+            catch (StackException)
+            {
+                Console.WriteLine("Incorrect expression");
+                throw;
```
Если Вы собираетесь выкинуть исключение, то незачем тут что-то писать на консоль --- выше по стеку вызовов его поймают и напишут, если надо. Вообще. на консоль должен писать только Main и специально предназначенные для этого классы.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+                Console.WriteLine("Incorrect expression");
+                throw;
+            }
+            if (inputString[i] == "+")
```
Это можно быть switch-ем сделать

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+            {
+                try
+                {
+                    if (secondNumber.CompareTo(0) == 0)
```
Почему так нельзя: https://docs.microsoft.com/en-us/dotnet/api/system.single.equals?view=net-6.0#system-single-equals(system-single)

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+                        throw new DivideByZeroException("division by 0");
+                    }
+                }
+                catch (DivideByZeroException exception)
```
:(

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,125 @@
+                {
+                    throw new InvalidCharacterException("Invalid character");
+                }
+                catch(InvalidCharacterException exception)
```
```suggestion
                catch (InvalidCharacterException exception)
```
Ну и да, так нельзя

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Solution.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace StackCalculator;
+
+public class Solution
+{
+    static void Main()
+    {
+        Console.WriteLine("Please, enter the expression");
+        var inputString = Console.ReadLine();
+        if (inputString == null)
+        {
+            return;
+        }
+        var subs = inputString.Split(' ');
+        Calculator stackCalculator = new Calculator();
+        Console.WriteLine($"{stackCalculator.CountTheExpressionInPostfixForm(subs)}");
+    }
+}
```
В C# 10 эти церемонии вообще не нужны:
```suggestion
﻿namespace StackCalculator;

Console.WriteLine("Please, enter the expression");
var inputString = Console.ReadLine();
if (inputString == null)
{
    return;
}
var subs = inputString.Split(' ');
Calculator stackCalculator = new Calculator();
Console.WriteLine($"{stackCalculator.CountTheExpressionInPostfixForm(subs)}");
```

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/StackCalculatorException.cs:
```diff
@@ -0,0 +1,17 @@
+/// <summary>
+/// A class for creating custom exceptions
+/// </summary>
+public class IncorrectExpressionException : Exception
```
В одном файле должен быть один класс, причём файл должен называться как класс, который в нём содержится

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,82 @@
+/// </summary>
+public class TestsStackCalculator
+{
+    Calculator? stackCalculator;
```
```suggestion
    private Calculator? stackCalculator;
```
В сообществе принято писать модификаторы видимости всегда, даже если умолчания устраивают

yurii-litvinov replied:
И чтобы не делать его nullable, можно инициализировать его тут же new Calculator();-ом, который потом, правда, перезатрётся в SetUp, но это не страшно

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,82 @@
+/// </summary>
+public class TestsStackCalculator
+{
+    Calculator? stackCalculator;
+    [SetUp]
```
```suggestion
    Calculator? stackCalculator;
    
    [SetUp]
```

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    [Test]
+    public void DevideByZero()
```
```suggestion
    public void DivideByZero()
```
Кстати, вот хорошая статья про именование юнит-тестов: https://dzone.com/articles/7-popular-unit-test-naming. Как у Вас --- плохо

yurii-litvinov commented on Stack/Stack/Stack.csproj:
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Не надо отдельного проекта для калькулятора и отдельного для стека, это как-то слишком мелкозернистое деление.

yurii-litvinov commented on Stack/Stack/Solution.cs:
```diff
@@ -0,0 +1,9 @@
+    static void Main()
+    {
+
+    }
```
Если содержательный Main не написать, собирайте проект как библиотеку (настраивается в свойствах проекта). Библиотека --- это то же самое, но без точки входа (то есть как раз без Main)

yurii-litvinov commented on Stack/Stack/IStack.cs:
```diff
@@ -0,0 +1,47 @@
+    /// <summary>
+    /// Function for removing an element from the stack
+    /// </summary>
+    /// <returns> Remote value</returns>
```
Remote --- это "удалённый" в смысле "на большой дистанции". Наверное, Removed?

yurii-litvinov commented on Stack/Stack/StackOnArray.cs:
```diff
@@ -0,0 +1,95 @@
+/// </summary>
+public class StackOnArray<T> : IStack<T>
+{
+    private T[]? values;
```
values инициализируется в конструкторе, зачем бы ему быть нуллабельным

yurii-litvinov commented on Stack/Stack/StackOnArray.cs:
```diff
@@ -0,0 +1,95 @@
+    {
+        if (values != null && numberOfElements == values.Length)
+        {
+            Array.Resize(ref values, values.Length + 20);
```
Лучше вдвое, это делает Push амортизированно константным. Сейчас Push амортизированно линейный.

yurii-litvinov commented on Stack/Stack/StackOnArray.cs:
```diff
@@ -0,0 +1,95 @@
+        {
+            throw new StackException("Stack is empty");
+        }
+        T topOfSTack = values[numberOfElements - 1];
+        numberOfElements--;
+        return topOfSTack;
```
Не-а, так в стеке остаётся ссылка на "удалённый" элемент, сборщик мусора не может его собрать, и это сродни утечке памяти (поскольку элемент так и будет сидеть в стеке до того, как его перезатрут новым).

yurii-litvinov commented on Stack/Stack/StackOnArray.cs:
```diff
@@ -0,0 +1,95 @@
+    /// <summary>
+    /// Function for removing the stack
+    /// </summary>
+    public override void DeleteStack() => values = null;
```
Для поддержания инварианта можно было сразу переинициализировать его новым массивом, и переименовать DeleteStack в ClearStack. Метод удаления стека вообще с концами тут  не нужен, мы просто можем "забыть" ссылку на сам стек.

yurii-litvinov commented on Stack/Stack/StackOnArray.cs:
```diff
@@ -0,0 +1,95 @@
+public class StackOnArray<T> : IStack<T>
+{
+    private T[]? values;
+    private int numberOfElements;
```
Правда ли, что numberOfElements всегда равно values.Length?

yurii-litvinov commented on Stack/Stack/StackOnLists.cs:
```diff
@@ -0,0 +1,109 @@
+/// <summary>
+/// A class for creating custom exceptions
+/// </summary>
+public class StackException : Exception
```
Каждый класс должен быть в своём файле

yurii-litvinov commented on Stack/Stack/StackOnLists.cs:
```diff
@@ -0,0 +1,109 @@
+{
+    private class StackElement
+    {
+        private StackElement? next { get; set; }
```
Свойства всегда называются с заглавной, а private-свойства вообще не нужны (они ничем не лучше полей)

yurii-litvinov commented on Stack/Stack/StackOnLists.cs:
```diff
@@ -0,0 +1,109 @@
+    private class StackElement
+    {
+        private StackElement? next { get; set; }
+        private T? value;
+        public T? Value { get => value; set { this.value = value; } }
```
Я говорил на паре так не писать :)

yurii-litvinov commented on Stack/Stack/StackOnLists.cs:
```diff
@@ -0,0 +1,109 @@
+    public override void Push(T value)
+    {
+        numberOfElements++;
+        StackElement newHead = new StackElement();
```
```suggestion
        StackElement newHead = new();
```

yurii-litvinov commented on Stack/Stack/StackOnLists.cs:
```diff
@@ -0,0 +1,109 @@
+            numberOfElements--;
+            return value;
+        }
+        return default(T);
```
```suggestion
        return default;
```

yurii-litvinov commented on Stack/Stack/StackOnLists.cs:
```diff
@@ -0,0 +1,109 @@
+    /// <returns>Top of the stack</returns>
+    public override T? ReturnTopOfTheStack()
+    {
+        if (head == null)
+        {
+            throw new StackException("Stack is empty");
+        }
+        return head == null ? default(T) : head.Value;
```
Ну...

yurii-litvinov commented on Stack/StackTest/StackTest.cs:
```diff
@@ -0,0 +1,103 @@
+/// </summary>
+public class StackTest
+{
+    private static IEnumerable<TestCaseData> Stacks
+    => new TestCaseData[]
+    {
+        new TestCaseData(new StackOnArray<int>()),
+        new TestCaseData(new StackOnLists<int>()),
+    };
```
```suggestion
    private static IEnumerable<TestCaseData> Stacks
        => new TestCaseData[]
        {
            new TestCaseData(new StackOnArray<int>()),
            new TestCaseData(new StackOnLists<int>()),
        };
```

yurii-litvinov commented on Stack/StackTest/StackTest.cs:
```diff
@@ -0,0 +1,103 @@
+    public void CheckTopOfTheStackAfterPush(IStack<int> stack)
+    {
+        stack?.Push(1);
+        Assert.AreEqual(stack?.ReturnTopOfTheStack(), 1);
```
Наоборот, сначала Expected, затем Actual. Иначе если тест не пройдёт, сообщение об ошибке будет лживым.

yurii-litvinov commented on Stack/StackTest/StackTest.cs:
```diff
@@ -0,0 +1,103 @@
+    [TestCaseSource(nameof(Stacks))]
+    public void CheckTopOfEmptyStack(IStack<int> stack)
+    {
+        var exception = Assert.Throws<StackException>(() => stack?.ReturnTopOfTheStack());
```
"?" тут не нужен, стек гарантированно не null

MinyazevR committed:
--Changed working with exceptions
--Each class in a separate file
--Changed tests
--The stack calculator class knows nothing about the stack implementation

MinyazevR committed:
Changed function names in tests

MinyazevR committed:
Removed unnecessary messages in exceptions

MinyazevR committed:
Removing unnecessary code

MinyazevR committed:
Changing the code by style guide

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStack.cs:
```diff
@@ -0,0 +1,76 @@
+        Assert.AreEqual(1, stack.NumberOfElements());
+        stack.Pop();
+        Assert.AreEqual(0, stack.NumberOfElements());
+    }
+
+
+    [TestCaseSource(nameof(Stacks))]
```
```suggestion
    }

    [TestCaseSource(nameof(Stacks))]
```

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,84 @@
+/// <summary>
+/// A class for testing a stack calculator
+/// </summary>
+public class TestsStackCalculator
```
```suggestion
public class StackCalculatorTests
```
Так каноничнее

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,84 @@
+    public void ShouldThrowsDivideByZeroExceptionWhenDivideByZero(IStack<float> stack)
+    {
+        string[] firstArray = { "123", "0", "/" };
+        Assert.Throws<System.DivideByZeroException>(() => stackCalculator?.CountTheExpressionInPostfixForm(firstArray, stack));
```
Вообще, калькулятор должен принимать строку, а не массив, но ладно :)

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,84 @@
+    public void ShouldThrowsInvalidCharacterExceptionWhenExpressionContainsInvalidCharacter(IStack<float> stack)
+    {
+        string[] firstArray = { "123", "0", "a" };
+        Assert.Throws<InvalidCharacterException> (() => stackCalculator?.CountTheExpressionInPostfixForm(firstArray, stack));
```
```suggestion
        Assert.Throws<InvalidCharacterException>(() => stackCalculator?.CountTheExpressionInPostfixForm(firstArray, stack));
```

yurii-litvinov commented on Homework2/StackCalculator/TestStackCalculator/TestStackCalculator.cs:
```diff
@@ -0,0 +1,84 @@
+    {
+        string[] expression = { "123", "4", "*", "-" };
+        Assert.Throws<IncorrectExpressionException>(() => stackCalculator?.CountTheExpressionInPostfixForm(expression, stack));
+    }
+
+
+    [TestCaseSource(nameof(Stacks))]
```
```suggestion
    }

    [TestCaseSource(nameof(Stacks))]
```

yurii-litvinov reviewed (approved):
Неаккуратно, но по существу всё ок, зачтена.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,90 @@
+    /// Function for counting expressions in postfix form
+    /// </summary>
+    /// <returns>Expression value</returns>
+    public float CountTheExpressionInPostfixForm(string[] inputString, IStack<float> stack)
```
Нет-нет, хочется, чтобы калькулятор принимал строку, а не массив строк

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/Calculator.cs:
```diff
@@ -0,0 +1,90 @@
+                secondNumber = stack.Pop();
+                firstNumber = stack.Pop();
+            }
+            catch (StackIsEmptyException)
+            {
+                throw;
+            }
```
Если вы думаете, что ваша жизнь бессмысленна, посмотрите на эти строки кода :)

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/IncorrectExpressionException.cs:
```diff
@@ -0,0 +1,9 @@
+/// </summary>
+public class IncorrectExpressionException : Exception
+{
+    public IncorrectExpressionException() : base() { }
```
Это тоже нечто ненужное. Если в классе не объявлен ни один конструктор, компилятор сам генерирует как раз это.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/StackOnArray.cs:
```diff
@@ -0,0 +1,99 @@
+    {
+        if (values != null && numberOfElements == values.Length)
+        {
+            Array.Resize(ref values, values.Length + 20);
```
Лучше увеличивать сразу вдвое. Иначе как у Вас операция Push работает в среднем за линию, а если стек растёт вдвое, то в среднем за константу.

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/StackOnArray.cs:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        numberOfElements++;
+        if (values == null)
```
Кажется, nullability-анализ гарантирует, что values не null, так что это довольно бессмысленная проверка

yurii-litvinov commented on Homework2/StackCalculator/StackCalculator/StackOnArray.cs:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        T topOfSTack = values[numberOfElements - 1];
+        Array.Clear(values, numberOfElements - 1, 1);
```
Это хитрый способ записать `values[numberOfElements - 1] = default;`

yurii-litvinov reviewed (changes requested):
Сам калькулятор со стеками ещё всё-таки надо немного поправить

MinyazevR committed:
add yml

MinyazevR committed:
fixed yml

MinyazevR committed:
rename yml
